crates/linker/src/pe.rs

```diff
@@ -1081,6 +1126,116 @@ fn write_section_header(
     data[section_header_start..][..header_array.len()].copy_from_slice(&header_array);
 }
 
+fn write_image_base_relocation(
+    mmap: &mut [u8],
+    reloc_section_start: usize,
+    new_block_va: u32,
+    relocations: &[u16],
+) -> usize {
+    // first, collect the blocks of relocations (each relocation block covers a 4K page)
+    // we will be moving stuff around, and have to work from the back to the front. However, the
+    // relocations are encoded in such a way that we can only decode them from front to back.
+    let mut next_block_start = reloc_section_start as u32;
+    let mut blocks = vec![];
+
+    loop {
+        let header =
+            load_struct_inplace_mut::<ImageBaseRelocation>(mmap, next_block_start as usize);
+
+        if header.virtual_address.get(LE) == 0 {
+            break;
+        }
+
+        blocks.push((next_block_start, *header));
+
+        next_block_start += header.size_of_block.get(LE);
+    }
+
+    // extra space that we'll use for the new relocations
+    let shift_amount = relocations.len() * 2;
+
+    // this is 8 in practice
+    const HEADER_WIDTH: usize = std::mem::size_of::<ImageBaseRelocation>();
+
+    let mut found_block = false;
+
+    // now, starting from the back, shift sections that need to be shifted and add the
+    // new relocations to the right block
+    while let Some((block_start, header)) = blocks.pop() {
+        let header_va = header.virtual_address.get(LE);
+        let header_size = header.size_of_block.get(LE);
+
+        let block_start = block_start as usize;
+
+        match header_va.cmp(&new_block_va) {
+            std::cmp::Ordering::Greater => {
+                // shift this block
+                mmap.copy_within(
+                    block_start..block_start + header_size as usize,
+                    block_start + shift_amount,
+                );
+            }
+            std::cmp::Ordering::Equal => {
+                // extend this block
+                let header = load_struct_inplace_mut::<ImageBaseRelocation>(mmap, block_start);
+
+                let new_size = header.size_of_block.get(LE) + shift_amount as u32;
+                header.size_of_block.set(LE, new_size);
+
+                let number_of_entries = (new_size as usize - HEADER_WIDTH) / 2;
```

Comment:
super nit: make 2 into a constant with a name. When reading this it took me a minute to realize that is the size of an entry.